### PR TITLE
 Add support for delayed lua evaluation and configuration edition

### PR DESCRIPTION
**Pull request by blinkseb**
A lazy value is an expression whose evaluation is delayed. In our context,
it essentially means a value which can be changed after the configuration
file is parsed.

An new lua function is added, `parameter` which accepts a string argument.
Instead of referencing directly the variable in the lua configuration, the
user must use this new function, and pass as argument the name of the
parameter, as specified in the `configuration` table (see updated lua
configuration for an example)

It's also now possible to edit parameters directly from the code instead of the configuration file. The new philosophy is now:

 1. You open a lua configuration file using `ConfigurationReader`
 2. You (may) edit the parameters you want
 3. You *freeze* the reader and receive a cloned version of the configuration that you can no longer edit.
 4. You use the frozen version of the configuration to create an instance of MoMEMta.
 5. Go back to step 2 if you want

--

TODO before merging:

 - [x] Ensure that a given parameter actually exists in the `configuration` table when evaluating the Lazy function (currently give a cryptic message)
 - [x] Support editing of a given parameter after the evaluation of the configuration file

### CODE REVIEW

**Comment by blinkseb** [comment]
I've updated the code a bit. It's now needed to `freeze` the configuration before passing it to MoMEMta. Next step is an interface to actually edit a parameter before freezing the configuration.

**Comment by blinkseb** [comment]
Pull-requested updated. On my side it seems to be complete. Editing of parameters work from the C++ code, but it's only limited to scalar type of the moment. It should be enough for the moment, and I hope it'd be rather easy to add the remaining type later.

File: core/include/lua/utils.h [hunk 1]
```diff
@@ -3,54 +3,263 @@
...
+     * to the top of the stack.
+     *
+     * \param L the current lua state
+     * \param value The value to convert
+     *
+     * \warning Vectors are currently not supported
+     */
+    void push_any(lua_State* L, const boost::any& value);
 
     template<typename T> T special_any_cast(const boost::any& value) {
```
Comment (swertz) [inline core/include/lua/utils.h]: Why is this needed?

Comment (blinkseb) [inline core/include/lua/utils.h]: It's used when parsing vectors, to provide implicit conversion from integer to double (see specialization of the template in the .cc file)

File: core/src/ConfigurationSet.cc [hunk 2]
```diff
@@ -48,7 +58,83 @@ void ConfigurationSet::parse(lua_State* L, int index) {
...
+
+    return lua::to_any(L, index);
+}
+
 bool ConfigurationSet::exists(const std::string& name) const {
     auto value = m_set.find(name);
     return (value != m_set.end());
 }
+
+void ConfigurationSet::setInternal(const std::string& name, Element& element, const boost::any& value) {
```
Comment (swertz) [inline core/src/ConfigurationSet.cc]: Why is the name needed here? It's also not used in LazyConfigurationSet...

File: core/src/ConfigurationSet.cc [hunk 3]
```diff
@@ -48,7 +58,83 @@ void ConfigurationSet::parse(lua_State* L, int index) {
...
+
+    element.value = value;
+    element.lazy = false;
+}
+
+void ConfigurationSet::freeze() {
+    if (frozen)
+        return;
+
+    frozen = true;
```
Comment (swertz) [inline core/src/ConfigurationSet.cc]: Maybe move this after the for loop, since it's not true if it gets interrupted (and someone catches the exceptions, who know...)?

File: core/src/ConfigurationSet.cc (hunk 2 above)
Comment (blinkseb) [inline core/src/ConfigurationSet.cc]: I thought it may be useful in the future in we add new subclasses

File: core/src/ConfigurationSet.cc (hunk 3 above)
Comment (blinkseb) [inline core/src/ConfigurationSet.cc]: If an exception is thrown during freezing, we would be half-frozen: some ConfigurationSet would be frozen, while some won't be. I honestly don't know how we can recover from such a state, so I don't think moving the frozen state after the for loop will change anything, but I can do it if you want :)

Comment (swertz) [inline core/src/ConfigurationSet.cc]: No it's fine ;)

**Comment by swertz** [comment]
Seems very good to me! A very, very big thanks for that!

Just a question before merging: why do you say it's limited to the scalar type? https://github.com/MoMEMta/MoMEMta/pull/28/files#diff-233043c4c174e0e51fc0ae56a94c4ceaR75 would also work with a vector right?

**Comment by vidalm** [comment]
I am lost, but I believe you !!!

**Comment by blinkseb** [comment]
Yes indeed, but the only thing the user can change are instances of `LazyConfigurationSet`, so this line: https://github.com/MoMEMta/MoMEMta/pull/28/files#diff-233043c4c174e0e51fc0ae56a94c4ceaR139

which call this method: https://github.com/MoMEMta/MoMEMta/pull/28/files#diff-d3b5c19707aaf6852add40ca17648b40R285

and there's no support for vector. It's possible to add, but it's a bit of work because you need to create the corresponding lua table from the vector.

**Comment by swertz** [comment]
Thanks for clearing that up!
